fem/transfer.cpp

```diff
@@ -522,6 +522,439 @@ const Operator &L2ProjectionGridTransfer::BackwardOperator()
+   // H refers to the higher-order mesh (DOFs correspond to columns)
+   M_LH = SparseMatrix(R->GetI(), R->GetJ(), NULL,
+                       R->Height(), R->Width(), false, true, true);
+
+   IntegrationPointTransformation ip_tr;
+   IsoparametricTransformation& emb_tr = ip_tr.Transf;
+
+   // Compute M_LH and R
+   for (int iho = 0; iho < nel_ho; ++iho)
+   {
+      Array<int> lor_els;
+      ho2lor.GetRow(iho, lor_els);
+      int nref = ho2lor.RowSize(iho);
+
+      Geometry::Type geom = mesh_ho->GetElementBaseGeometry(iho);
+      const FiniteElement& fe_ho = *fes_ho.GetFE(iho);
+      const FiniteElement& fe_lor = *fes_lor.GetFE(lor_els[0]);
+      int nedof_ho = fe_ho.GetDof();
+      int nedof_lor = fe_lor.GetDof();
+
+      Vector shape_ho(nedof_ho);
+      Vector shape_lor(nedof_lor);
+      Array<int> dofs_ho(nedof_ho);
+      Array<int> dofs_lor(nedof_lor);
+
+      emb_tr.SetIdentityTransformation(geom);
+      const DenseTensor& pmats = cf_tr.point_matrices[geom];
+
+      DenseMatrix M_LH_el(nedof_lor, nedof_ho);
+      DenseMatrix R_el(nedof_lor, nedof_ho);
+
+      for (int iref = 0; iref < nref; ++iref)
+      {
+         int ilor = lor_els[iref];
+         ElementTransformation* el_tr = fes_lor.GetElementTransformation(ilor);
+
+         // Create the transformation that embeds the fine low-order element
+         // within the coarse high-order element in reference space
+         emb_tr.SetPointMat(pmats(cf_tr.embeddings[ilor].matrix));
+
+         int order = fe_lor.GetOrder() + fe_ho.GetOrder() + el_tr->OrderW();
+         const IntegrationRule* ir = &IntRules.Get(geom, order);
+         M_LH_el = 0.0;
+         for (int i = 0; i < ir->GetNPoints(); i++)
+         {
+            const IntegrationPoint& ip_lor = ir->IntPoint(i);
+            IntegrationPoint ip_ho;
+            ip_tr.Transform(ip_lor, ip_ho);
+            fe_lor.CalcShape(ip_lor, shape_lor);
+            fe_ho.CalcShape(ip_ho, shape_ho);
+            el_tr->SetIntPoint(&ip_lor);
+            // For now we use the geometry information from the LOR space
+            // which means we won't be mass conservative if the mesh is curved
+            double w = el_tr->Weight() * ip_lor.weight;
+            shape_lor *= w;
+            AddMultVWt(shape_lor, shape_ho, M_LH_el);
+         }
+         fes_lor.GetElementDofs(ilor, dofs_lor);
+         Vector R_row;
+         for (int i = 0; i < nedof_lor; ++i)
+         {
+            M_LH_el.GetRow(i, R_row);
+            R_el.SetRow(i, R_row.Set(ML_inv[dofs_lor[i]], R_row));
+         }
+         fes_ho.GetElementDofs(iho, dofs_ho);
+         M_LH.AddSubMatrix(dofs_lor, dofs_ho, M_LH_el);
+         R->AddSubMatrix(dofs_lor, dofs_ho, R_el);
+      }
+   }
+}
+
+L2ProjectionH1GridTransfer::L2ProjectionH1::~L2ProjectionH1()
+{
+   delete R;
+   if (P) { delete P; }
+}
+
+void L2ProjectionH1GridTransfer::L2ProjectionH1::Mult(
+   const Vector& x, Vector& y) const
+{
+   int vdim = fes_ho.GetVDim();
```

Comment:
I thought about ways to do this (either virtual functions or a policy-based design).  I agree it's nice to reduce code reproduction, but I'm worried in this case, since the code is relatively short and the number of functions that would utilize this abstraction is small (4), that introducing more general code might not be worth the increased complexity and loss of readability the abstracted code would introduce.  However, I'd be happy to implement something to reduce some of the code redundancy.  Did you have a particular implementation in mind?

Reply:
I think it already looks a lot better now that you've added `FiniteElementSpace::GetVDofs`.

Reply:
Great!  I'll leave it as is.